platform/x11/cog-platform-x11.c

```diff
@@ -0,0 +1,896 @@
+    glEnableVertexAttribArray (gl_data.attr_texture);
+
+    glDrawArrays (GL_TRIANGLE_STRIP, 0, 4);
+
+    glDisableVertexAttribArray (gl_data.attr_pos);
+    glDisableVertexAttribArray (gl_data.attr_texture);
+
+    eglSwapBuffers (egl_data.display, egl_data.surface);
+
+    wpe_view_data.image = image;
+    xcb_data.needs_initial_paint = false;
+    xcb_data.needs_frame_completion = true;
+    xcb_schedule_repaint ();
+}
+
+static void
+xcb_frame_completion (void)
+{
+    if (wpe_view_data.image) {
+        wpe_view_backend_exportable_fdo_egl_dispatch_release_exported_image (wpe_host_data.exportable, wpe_view_data.image);
+        wpe_view_data.image = NULL;
+    }
+
+    wpe_view_backend_exportable_fdo_dispatch_frame_complete (wpe_host_data.exportable);
+}
+
+static void
+xcb_handle_key_press (xcb_key_press_event_t *event)
+{
+    uint32_t keysym = xkb_state_key_get_one_sym (xkb_data.state, event->detail);
+    uint32_t unicode = xkb_state_key_get_utf32 (xkb_data.state, event->detail);
+
+    struct wpe_input_keyboard_event input_event = {
+        event->time,
+        keysym,
+        unicode,
+        true,
+        xkb_data.modifiers
+    };
+    wpe_view_backend_dispatch_keyboard_event (wpe_view_data.backend, &input_event);
+}
+
+static void
+xcb_handle_key_release (xcb_key_press_event_t *event)
+{
+    uint32_t keysym = xkb_state_key_get_one_sym (xkb_data.state, event->detail);
+    uint32_t unicode = xkb_state_key_get_utf32 (xkb_data.state, event->detail);
+
+    struct wpe_input_keyboard_event input_event = {
+        event->time,
+        keysym,
+        unicode,
+        false,
+        xkb_data.modifiers
+    };
+    wpe_view_backend_dispatch_keyboard_event (wpe_view_data.backend, &input_event);
+}
+
+static void
+xcb_handle_axis (xcb_button_press_event_t *event, int16_t axis_delta[2])
+{
+#if HAVE_2D_AXIS_EVENT
+    struct wpe_input_axis_2d_event input_event = { 0, };
+    input_event.base.type = wpe_input_axis_event_type_mask_2d | wpe_input_axis_event_type_motion_smooth;
+    input_event.base.time = event->time;
+    input_event.base.x = xcb_data.pointer.x;
+    input_event.base.y = xcb_data.pointer.y;
+
+    input_event.x_axis = axis_delta[0];
+    input_event.y_axis = axis_delta[1];
+
+    wpe_view_backend_dispatch_axis_event (wpe_view_data.backend, &input_event.base);
+#else
+    assert (axis_delta[0] ^ axis_delta[1]);
+
+    struct wpe_input_axis_event input_event = {
+        wpe_input_axis_event_type_motion,
+        event->time,
+        xcb_data.pointer.x, xcb_data.pointer.y,
+        0, 0, 0,
+    };
+
+    if (!!axis_delta[0]) {
+        input_event.axis = 1;
+        input_event.value = axis_delta[0];
+    } else {
+        input_event.axis = 0;
+        input_event.value = axis_delta[1];
+    }
+
+    wpe_view_backend_dispatch_axis_event (wpe_view_data.backend, &input_event);
+#endif
+}
+
+static void
+xcb_handle_button_press (xcb_button_press_event_t *event)
+{
+    int16_t axis_delta[4][2] = {
```

Comment:
This is a read-only lookup table, it would be good to mark it `static const` to help the compiler optimize it away.